### ISSUE TITLE
Continuous upload of incremental sstables

### ISSUE THREAD

**Issue by eriksw** [thelastpickle/cassandra-medusa#14]
Is there any intention for Medusa to support a continuous/semi-continuous operating mode where, with Cassandra configured to write incremental backups, that Medusa will upload (and clean) them all? (And on the restore side, restore the latest full snapshot then the available incremental tables from after then?)

I'm trying to determine the path of least resistance for our needs: we need to minimize the window of data lost in the event of a total cluster loss. Daily or even hourly snapshots are not good enough.

We have [a tool](https://github.com/retailnext/cassandrabackup) that does this, but it is lacking in a few ways (pruning, cloud portability, general code fixups/test coverage), so I'm trying to get a feeling for whether extending/improving Medusa is a viable (/better) option than finishing our tool.

**Comment by adejanovski** [comment]
Yes, that's a feature we'd really like to add, which would allow Medusa to perform point in time restores.
We'd love to help you contribute such a feature to Medusa and work on the design with you.

**Comment by nicholasamorim** [comment]
What does this mean? If someone explains it further I can try to contribute as well.

Would this be something like running in daemon mode and using something such as `inotify` to act straight away everytime a SSTtable is written?

**Comment by eriksw** [comment]
I don't know if this is how medusa should do this, but I'll describe what the tool I wrote does and some key lessons learned:

**There are lot of gotchas with inotify and the consequences of missing a file are extremely severe, so I don't use it at all and just do a scan on a frequent basis for incremental sstables.**

There's multiple kinds of manifests (snapshot, incremental, incomplete). Manifests are stored in per-node folders within the storage bucket, **named with the timestamp and manifest type all in the filename**.

The timestamp of the manifest is, for a snapshot the moment when nodetool was invoked to create it, and for incremental the moment when the file walk began. **I complete the filesystem walk building a list of files as quickly as possible to make the incremental manifest atomic-ish.** (Even if it takes a long time to upload the files, the margin by which a newer sstable might be in an older-timestamped manifest is about a second. This is important for "I need these CFs from before bad thing happened at time X"-type disaster recovery.)

Restoring is on a per-node basis with my tool, and "not before" and "not after" flags are available. **The _last_ "full snapshot" manifest found for the node within the period is selected, and merged with all the incremental/incomplete manifests after it** (again bounded by flags) to produce the list of blobs to restore.

(Every node does periodic incremental (5m) and snapshot (60m) backups on its own with no coordination; restore coordination is deferred parallel SSH-equivalent tools.)

The tool I wrote relies heavily on caching. The file storage in the bucket is content-addressed; there is caching of what blobs are known to be in the bucket (and how long they're [object locked](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock.html) for), as well of caching of inode -> ((mtime, size), digests). This cache is persistent. Unless the cache is wiped out, my backup tool will read a given file _at most twice_: once to calculate overall and part digests (for multipart upload), another to upload it to the bucket if it's not in the blob store.

**The work for a "full snapshot" backup is little more than an incremental one, thanks to the caches being hot from the frequent incremental processing.** (It takes a bit of time for cassandra to make/remove all the hard links, and the manifest upload is a bit bigger, but that's it.)

**Frequent "full" snapshots is important to bound the effects of write amplification.** When restoring 1:1 instead of streaming into sstableloader or whatever: _write amplification * time = restore space requirement amplification_. (The space required on each node is exactly equivalent to the snapshot plus if you let incrementals pile up for the entire time since the last snapshot, plus room for compaction of all that.)
